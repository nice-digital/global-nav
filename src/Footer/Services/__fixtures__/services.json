{
  "external": [
    {
      "host": "www.test.nice.org",
      "landingPath": "/url1/",
      "text": "First link",
      "id": "link1",
      "footer": true
    },
    {
      "host": "www.test.nice.org",
      "landingPath": "/url2/",
      "text": "Second link",
      "id": "link2",
      "footer": false
    },
    {
      "host": "www.test.nice.org",
      "landingPath": "/url3/",
      "text": "Third link",
      "id": "link3",
      "footer": true
    }
  ]
}
