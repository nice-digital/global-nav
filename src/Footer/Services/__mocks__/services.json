{
  "external": [
    {
      "landingUrl": "https://url1/",
      "text": "First link",
      "id": "link1",
      "footer": true
    },
    {
      "landingUrl": "https://url2/",
      "text": "Second link",
      "id": "link2",
      "footer": false
    },
    {
      "landingUrl": "https://url3/",
      "text": "Third link",
      "id": "link3",
      "footer": true
    }
  ]
}
