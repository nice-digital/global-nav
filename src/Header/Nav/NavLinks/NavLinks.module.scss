/* stylelint-disable no-descending-specificity */

@import '../../header-settings';
@import '../../../Common.module.scss';
@import '../Nav.module.scss';

.menuList {
  @extend .menuList;
}

a.link,
button.link {
  @extend .linkStyling;
}

.dropdown {
  background-color: $nds-colour-nice-grey-3;
  box-shadow: 0 rem($nds-spacing-small) rem($nds-spacing-small) 0
    rgb(0 0 0 / 10%);
  display: none;
  left: 0;
  position: absolute;
  width: 100%;
  z-index: 1;

  &.active {
    display: block;
  }
}
