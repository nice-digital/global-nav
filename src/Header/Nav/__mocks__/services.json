{
  "rootUrl": "https://rootUrl.org.uk",
  "external": [
    {
      "href": "https://url1/",
      "text": "First link",
      "id": "link1",
      "dropdown": true,
      "dropdownComponent": "FirstComponent"
    },
    {
      "href": "https://url2/",
      "text": "Second link",
      "id": "link2",
      "dropdown": false,
      "abbreviation": true,
      "title": "Abbreviation title",
      "links": [
        {
          "text": "Sub link 1",
          "href": "/"
        },
        {
          "text": "Sub link 2",
          "href": "/test"
        }
      ]
    },
    {
      "href": "https://url3/",
      "text": "Third link",
      "id": "link3",
      "dropdown": true,
      "dropdownComponent": "ThirdComponent",
      "abbreviation": true,
      "title": "Abbreviation title",
      "links": [
        {
          "text": "Sub link 1",
          "href": "/"
        },
        {
          "text": "Sub link 2",
          "href": "/test"
        }
      ],
      "nestedLinks": [
        {
          "href": "https://nestedLink1",
          "text": "Nested link one",
          "id": "nested-1"
        },
        {
          "href": "https://nestedLink2",
          "text": "Nested link two",
          "id": "nested-2"
        }
      ]
    }
  ],
  "internal": [
    {
      "href": "/",
      "text": "First internal service",
      "id": "internal1",
      "links": [
        {
          "text": "internal1 sub link 1",
          "href": "/i1sublink1"
        },
        {
          "text": "internal1 sub link 2",
          "href": "/i2sublink2"
        }
      ]
    },
    {
      "href": "/",
      "text": "Second internal service",
      "id": "internal2",
      "links": [
        {
          "text": "internal2 sub link 1",
          "href": "/i2sublink1"
        },
        {
          "text": "internal2 sub link 2",
          "href": "/i2sublink2"
        }
      ]
    }
  ]
}
