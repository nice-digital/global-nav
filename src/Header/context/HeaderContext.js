import React, { useEffect, useState, createContext } from "react";
import PropTypes from "prop-types";

const defaultValues = { accountMenuIsExpanded: false, idOfOpenDropdown: null };

export const HeaderContext = createContext(defaultValues);

export const HeaderContextProvider = function ({ children }) {
	const [idOfOpenDropdown, setidOfOpenDropdown] = useState(null);
	const [accountMenuIsExpanded, setAccountMenuIsExpanded] = useState(false);

	const value = {
		idOfOpenDropdown,
		setidOfOpenDropdown,
		accountMenuIsExpanded,
		setAccountMenuIsExpanded,
	};

	const locationOrHashChangeHandler = () => {
		console.log("closing any open dropdown due to location or hash change");
		setidOfOpenDropdown(null);
	};

	useEffect(() => {
		window.addEventListener("hashchange", locationOrHashChangeHandler);
		return () => {
			window.removeEventListener("hashchange", locationOrHashChangeHandler);
		};
	}, []);

	useEffect(() => {
		locationOrHashChangeHandler();
	}, [location]);

	useEffect(() => {
		// we know both are now open
		if (accountMenuIsExpanded && idOfOpenDropdown) {
			setAccountMenuIsExpanded((prevState) => {
				// if prevstate is not null, you know it was already open, therefore close it
				if (prevState) return false;
			});
		}
	}, [accountMenuIsExpanded, idOfOpenDropdown]);

	return (
		<HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
	);
};

HeaderContextProvider.propTypes = {
	children: PropTypes.node,
};
