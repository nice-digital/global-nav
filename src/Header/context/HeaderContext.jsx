import {
	useEffect,
	useState,
	createContext,
	createRef,
	useCallback,
} from "react";

import PropTypes from "prop-types";

const defaultValues = { accountMenuIsExpanded: false, idOfOpenDropdown: null };

export const HeaderContext = createContext(defaultValues);

export const HeaderContextProvider = function ({ children }) {
	const [idOfOpenDropdown, setidOfOpenDropdown] = useState(null);
	const [accountMenuIsExpanded, setAccountMenuIsExpanded] = useState(false);
	const clickOutsideRef = createRef();

	const value = {
		idOfOpenDropdown,
		setidOfOpenDropdown,
		accountMenuIsExpanded,
		setAccountMenuIsExpanded,
		clickOutsideRef,
	};

	const handleClickOutsideHeaderMenu = useCallback(
		(e) => {
			if (
				clickOutsideRef.current &&
				clickOutsideRef.current.contains(e.target) == false
			) {
				setidOfOpenDropdown(null);
			}
		},
		[clickOutsideRef]
	);

	useEffect(() => {
		// listen for clicks outside of header menu to close dropdown
		document.body.addEventListener("click", handleClickOutsideHeaderMenu);
		return () => {
			document.body.removeEventListener("click", handleClickOutsideHeaderMenu);
		};
	}, [handleClickOutsideHeaderMenu]);

	const locationHref = typeof location == "undefined" ? null : location.href;

	useEffect(() => {
		setidOfOpenDropdown(null);
	}, [locationHref]);

	const popStateHandler = () => {
		setidOfOpenDropdown(null);
	};

	useEffect(() => {
		window && window.addEventListener("popstate", popStateHandler);
		return () => {
			window && window.removeEventListener("popstate", popStateHandler);
		};
	}, []);

	useEffect(() => {
		// we know both are now open
		if (accountMenuIsExpanded && idOfOpenDropdown) {
			setAccountMenuIsExpanded((prevState) => {
				// if prevstate is not null, you know it was already open, therefore close it
				if (prevState) return false;
			});
		}
	}, [accountMenuIsExpanded, idOfOpenDropdown]);

	return (
		<HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
	);
};

HeaderContextProvider.propTypes = {
	children: PropTypes.node,
};
