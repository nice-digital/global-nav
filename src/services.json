{
  "external": [
    {
      "href": "https://pathways.nice.org.uk/",
      "text": "NICE Pathways",
      "id": "pathways",
      "links": [
        {
          "text": "Browse",
          "href": "/#browse"
        },
        {
          "text": "Recently viewed",
          "href": "/#recently-viewed"
        },
        {
          "text": "Help",
          "href": "/#help"
        },
        {
          "text": "Leave feedback",
          "href": "https://www.nice.org.uk/NICE-Pathways-feedback"
        }
      ]
    },
    {
      "href": "https://www.nice.org.uk/guidance",
      "text": "NICE guidance",
      "id": "guidance"
    },
    {
      "href": "https://www.nice.org.uk/about/what-we-do/life-sciences",
      "text": "Life sciences",
      "id": "life-sciences"
    },
    {
      "href": "https://www.nice.org.uk/standards-and-indicators",
      "text": "Standards and indicators",
      "id": "standards"
    },
    {
      "href": "https://www.evidence.nhs.uk/",
      "text": "Evidence search",
      "id": "evidence"
    },
    {
      "href": "https://bnf.nice.org.uk/",
      "text": "BNF",
      "abbreviation": true,
      "title": "British National Formulary",
      "id": "bnf",
      "links": [
        {
          "text": "Drugs",
          "href": "/drug/"
        },
        {
          "text": "Interactions",
          "href": "/interaction/"
        },
        {
          "text": "Treatment summaries",
          "href": "/treatment-summary/"
        },
        {
          "text": "What's changed?",
          "href": "/about/changes.html"
        },
        {
          "text": "About BNF",
          "href": "/about/"
        }
      ]
    },
    {
      "href": "https://bnfc.nice.org.uk/",
      "text": "BNFC",
      "abbreviation": true,
      "title": "British National Formulary for Children",
      "id": "bnfc",
      "links": [
        {
          "text": "Drugs",
          "href": "/drug/"
        },
        {
          "text": "Interactions",
          "href": "/interaction/"
        },
        {
          "text": "Treatment summaries",
          "href": "/treatment-summary/"
        },
        {
          "text": "What's changed?",
          "href": "/about/changes.html"
        },
        {
          "text": "About BNFC",
          "href": "/about/"
        }
      ]
    },
    {
      "href": "https://cks.nice.org.uk/",
      "text": "CKS",
      "abbreviation": true,
      "title": "Clinical Knowledge Summaries",
      "id": "cks",
      "links": [
        {
          "image": "cks",
          "text": "CKS - Home",
          "href": "/"
        },
        {
          "text": "Health topics A to Z",
          "href": "/topics/"
        },
        {
          "text": "Specialities",
          "href": "/specialities/"
        },
        {
          "text": "What's new",
          "href": "/whats-new/"
        },
        {
          "text": "About CKS",
          "href": "/about/"
        }
      ]
    },
    {
      "href": "https://www.nice.org.uk/about/what-we-do/evidence-services/journals-and-databases",
      "text": "Journals and databases",
      "id": "journals"
    }
  ],
  "internal": [
    {
      "href": "/",
      "text": "NICE In development",
      "id": "indev",
      "links": [
        {
          "text": "Guidance projects",
          "href": "/"
        },
        {
          "text": "Advice projects",
          "href": "/adviceprojects"
        }
      ]
    }
  ]
}
