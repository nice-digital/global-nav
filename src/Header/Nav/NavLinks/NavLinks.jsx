import React, { useEffect } from "react";
import classnames from "classnames";
import FocusTrap from "focus-trap-react";
import PropTypes from "prop-types";
import useEventListener from "@use-it/event-listener";
import ChevronDown from "@nice-digital/icons/lib/ChevronDown";
import ChevronUp from "@nice-digital/icons/lib/ChevronUp";
import SubNav from "../SubNav";
import Dropdown from "../Dropdown";
import { HeaderContext } from "../../context/HeaderContext";

import styles from "./NavLinks.module.scss";
import {
	trackEvent,
	defaultEventCategory,
	headerClickEventAction,
} from "../../../tracker";

export function NavLinks({
	servicesToDisplay,
	currentService,
	subLinks,
	onNavigating,
	skipLinkId,
}) {
	const [canUseDOM, setCanUseDOM] = React.useState(false);
	const { idOfOpenDropdown, setidOfOpenDropdown } =
		React.useContext(HeaderContext);

	useEffect(() => {
		setCanUseDOM(true);
	}, []);

	const ESCAPE_KEYS = ["27", "Escape", "Esc"];

	function handleNavButtonClick(id) {
		setidOfOpenDropdown(id === idOfOpenDropdown ? null : id);
	}

	function handleNavLinkClick(e) {
		e.preventDefault();
		setidOfOpenDropdown(null);

		const href = e.currentTarget.getAttribute("href");
		trackEvent(
			defaultEventCategory,
			headerClickEventAction,
			e.currentTarget.textContent,
			null,
			(function () {
				window.location.href = href;
			})()
		);
	}

	function escapeDropdown({ key }) {
		if (ESCAPE_KEYS.includes(String(key))) setidOfOpenDropdown(null);
	}

	useEventListener("keydown", escapeDropdown);

	const options = {
		clickOutsideDeactivates: true,
		initialFocus: false,
	};

	return (
		<FocusTrap active={idOfOpenDropdown !== null} focusTrapOptions={options}>
			<ul className={styles.menuList} aria-labelledby="header-menu-button">
				{servicesToDisplay.map(
					(
						{
							href,
							id,
							text,
							abbreviation,
							title,
							dropdown,
							dropdownComponent,
							nestedLinks,
						},
						index
					) => {
						let ariaCurrent = null;

						if (currentService && id === currentService) {
							ariaCurrent = true;

							if (
								typeof location !== "undefined" &&
								location &&
								href ===
									`${location.protocol}//${location.host}${location.pathname}`
							) {
								ariaCurrent = "page";
							}
						}

						return (
							<li key={id} data-tracking={text}>
								{dropdown && canUseDOM ? (
									<button
										onClick={() => handleNavButtonClick(id)}
										aria-current={ariaCurrent}
										className={styles.link}
										aria-controls={`dropdown-${id}`}
										aria-expanded={id === idOfOpenDropdown ? true : false}
										id={`navlink-${id}`}
										aria-label={abbreviation && title}
									>
										<span>{text}</span>
										{id === idOfOpenDropdown ? (
											<ChevronUp className={styles.icon} pointerEvents="none" />
										) : (
											<>
												<ChevronDown
													className={styles.icon}
													pointerEvents="none"
												/>
											</>
										)}
									</button>
								) : (
									<>
										<a
											href={href}
											aria-current={ariaCurrent}
											className={styles.link}
											onClick={handleNavLinkClick}
											id={`navlink-${id}`}
										>
											<span aria-label={abbreviation && title}>{text}</span>
											{nestedLinks && (
												<ChevronDown
													className={styles.icon}
													pointerEvents="none"
												/>
											)}
										</a>

										{nestedLinks && (
											<>
												<ul
													id={id}
													className={styles.nonJsDropdown}
													aria-label="More NICE services"
												>
													{nestedLinks.map(({ href, text, id }) => {
														return (
															<li key={id}>
																<a href={href}>{text}</a>
															</li>
														);
													})}
												</ul>
											</>
										)}
									</>
								)}
								{dropdown && canUseDOM ? (
									<Dropdown
										component={dropdownComponent}
										className={classnames([
											styles.dropdown,
											id === idOfOpenDropdown && styles.active,
										])}
										text={text}
										nextNavSlug={
											servicesToDisplay[index + 1]
												? servicesToDisplay[index + 1]["id"]
												: skipLinkId
										}
										closeDropdown={() => setidOfOpenDropdown(null)}
										id={`dropdown-${id}`}
									/>
								) : null}
								{ariaCurrent && subLinks && (
									<SubNav
										links={subLinks}
										text={text}
										onNavigating={onNavigating}
									/>
								)}
							</li>
						);
					}
				)}
			</ul>
		</FocusTrap>
	);
}

NavLinks.propTypes = {
	skipLinkId: PropTypes.string,
	servicesToDisplay: PropTypes.array,
	currentService: PropTypes.string,
	subLinks: PropTypes.array,
	onNavigating: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
	handleScrim: PropTypes.func,
};

export default NavLinks;
