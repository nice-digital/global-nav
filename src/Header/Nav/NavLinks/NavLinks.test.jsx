import React from "react";
import { NavLinks } from "./NavLinks";
import { createEvent, fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import services from "./../__mocks__/services.json";
import { HeaderContextProvider } from "../../context/HeaderContext";

describe("NavLinks", () => {
	const defaultProps = {
		servicesToDisplay: services.external,
	};

	it.each(
		services.external
			.filter(({ header }) => header)
			.map(({ text }) => ({ text }))
	)("should render menu item for $text", ({ text }) => {
		const { getByText } = render(<NavLinks {...defaultProps} />);
		expect(getByText(text)).toBeInTheDocument();
	});

	it("should render a top level button if there's a dropdown component", () => {
		const { getByRole } = render(<NavLinks {...defaultProps} />);
		expect(getByRole("button", { name: "First link" })).toBeInTheDocument();
	});

	it("should render a top level anchor if there's no dropdown", () => {
		const { getByRole } = render(<NavLinks {...defaultProps} />);
		expect(
			getByRole("link", { name: "Second link abbreviation title" })
		).toBeInTheDocument();
	});

	it("should add aria-expanded=true to the button of the currently expanded dropdown", async () => {
		const { getByRole } = render(
				<HeaderContextProvider>
					<NavLinks {...defaultProps} />
				</HeaderContextProvider>
			),
			button = getByRole("button", { name: "First link" }),
			user = userEvent.setup();

		await user.click(button);

		expect(button).toHaveAttribute("aria-expanded", "true");
	});

	it("should add aria-current=true if the global config currentService matches current service", async () => {
		const { getByRole } = render(
				<HeaderContextProvider>
					<NavLinks
						{...defaultProps}
						currentService={services.external[0].id}
					/>
				</HeaderContextProvider>
			),
			button = getByRole("button", { name: "First link" }),
			user = userEvent.setup();

		await user.click(button);

		expect(button).toHaveAttribute("aria-current", "true");
	});

	it("should add an abbreviation title label", () => {
		const { getByRole } = render(<NavLinks {...defaultProps} />),
			secondLink = getByRole("link", {
				name: "Second link abbreviation title",
			});

		expect(secondLink).toHaveTextContent("Second link");
	});

	it("should prevent default and navigate in event callback on nav item click", () => {
		const { getByRole } = render(
				<HeaderContextProvider>
					<NavLinks {...defaultProps} currentService="link2" />
				</HeaderContextProvider>
			),
			secondLink = getByRole("link", {
				name: "Second link abbreviation title",
			}),
			clickEvent = createEvent.click(secondLink);

		fireEvent(secondLink, clickEvent);

		expect(clickEvent.defaultPrevented).toBeTruthy();

		window.dataLayer[0].eventCallback();
		expect(window.location).toBeAt("https://www.test-link2.nice.org/url2/");
	});

	it("should push dataLayer event for nav item click", async () => {
		const { getByRole } = render(
				<HeaderContextProvider>
					<NavLinks {...defaultProps} currentService="link2" />
				</HeaderContextProvider>
			),
			secondLink = getByRole("link", {
				name: "Second link abbreviation title",
			}),
			user = userEvent.setup();

		await user.click(secondLink);

		expect(window.dataLayer).toStrictEqual([
			{
				destinationUrl: "https://www.test-link2.nice.org/url2/",
				event: "GlobalNav",
				eventCategory: "TopHat and footer",
				eventAction: "Tophat click",
				eventLabel: "Second link",
				eventCallback: expect.any(Function),
				eventTimeout: 2000,
			},
		]);
	});

	it("Matches snapshot with sub links for selected external service", () => {
		const externalServices = services.external;

		const { container } = render(
			<NavLinks {...defaultProps} service={externalServices[1].id} />
		);
		expect(container).toMatchSnapshot();
	});
});
