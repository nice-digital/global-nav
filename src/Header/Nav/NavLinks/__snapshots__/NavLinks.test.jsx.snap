// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavLinks Matches snapshot with sub links for selected external service 1`] = `
<FocusTrap
  _createFocusTrap={[Function]}
  active={false}
  focusTrapOptions={
    Object {
      "clickOutsideDeactivates": true,
      "initialFocus": false,
    }
  }
  paused={false}
>
  <ul
    aria-labelledby="header-menu-button"
    className="menuList"
  >
    <li
      data-tracking="First link"
      key="link1"
    >
      <button
        aria-controls="dropdown-link1"
        aria-current={null}
        aria-expanded={false}
        className="link"
        id="navlink-link1"
        onClick={[Function]}
      >
        <span
          id="link1"
        >
          First link
        </span>
        <SvgChevronDown
          className="icon"
          pointerEvents="none"
        />
      </button>
      <Dropdown
        className="dropdown"
        closeDropdown={[Function]}
        component="FirstComponent"
        id="dropdown-link1"
        nextNavSlug="link2"
        text="First link"
      />
    </li>
    <li
      data-tracking="Second link"
      key="link2"
    >
      <a
        aria-current={null}
        className="link"
        href="https://url2/"
        id="navlink-link2"
        onClick={[Function]}
      >
        <span
          aria-label="Abbreviation title"
          id="link2"
        >
          Second link
        </span>
      </a>
    </li>
    <li
      data-tracking="Third link"
      key="link3"
    >
      <button
        aria-controls="dropdown-link3"
        aria-current={null}
        aria-expanded={false}
        className="link"
        id="navlink-link3"
        onClick={[Function]}
      >
        <span
          aria-label="Abbreviation title"
          id="link3"
        >
          Third link
        </span>
        <SvgChevronDown
          className="icon"
          pointerEvents="none"
        />
      </button>
      <Dropdown
        className="dropdown"
        closeDropdown={[Function]}
        component="ThirdComponent"
        id="dropdown-link3"
        text="Third link"
      />
    </li>
  </ul>
</FocusTrap>
`;
